### PR TITLE
Code object compression via bundling

### PR DESCRIPTION
**Pull request by bstefanuk**
**Summary:**

This PR adds a compression layer to all final code objects, thereby generating smaller libraries at the expense of build time. Includes minor refactoring.

**Outcomes:**

- A new clang-offload-bundler invocation is added after assembly object linking.
- `getAssemblyCodeObjectFiles` has been renamed to `buildAssemblyCodeObjectFiles` to match the name of source kernel functions.
- The flag `--no-compress` can be used on the install script to disable compression.

| Build | Time   | Build Size (*build/library/* directory)                  |
|------------------|---------------|--------------------------------|
| **TCL, Feature, gfx90a** |   8m38.902s (8.65s) | 269M    |
| **TCL, Develop, gfx90a** |   8m3.921s (8.07s)  | 812M    |
| **hipBLASLt, Feature, gfx942** | 59m9.664s (59.16s) |   1.2G |
| **hipBLASLt, Develop, gfx942** | 56m52.821s (56.88s) |  11G  |

Build time increase, gfx90a: 7.0%
Build time increase for hipBLASLt, gfx942: 4.0%

Compression ratio gfx90a: 3.02
Compression ratio for hipBLASLt gfx942 libraries: 8.99

**Testing and Environment:**

Docker: Ubuntu 24.04, ROCm 6.4 RC stack, AMD clang version 18.0.0, AMD clang-offload-bundler version 18.0.0

Tested with hipBLASLt test client
```
[==========] 14429 tests from 13 test suites ran. (1977818 ms total)
[  PASSED  ] 14429 tests.
```

### CODE REVIEW

**Comment by bstefanuk** [comment]
Once documentation and tests are added, I will remove from draft.

File: tensilelite/Tensile/TensileCreateLibrary.py
Comment (bstefanuk) [inline tensilelite/Tensile/TensileCreateLibrary.py]: This is more appropriately named `numBatches`

File: tensilelite/Tensile/TensileCreateLibrary.py
Comment (TorreZuk) [inline tensilelite/Tensile/TensileCreateLibrary.py]: Can you consider doing what I added for windows (response file) instead of processing in batches?    Is there a limit of 10000 with the linker?

**Review by TorreZuk — DISMISSED** [review]
Basics look good, assume all rocblas tests passed.

File: tensilelite/Tensile/TensileCreateLibrary.py
Comment (bstefanuk) [reply tensilelite/Tensile/TensileCreateLibrary.py]: In TensileLite, this batching strategy is implemented, but the source PR doesn't describe the motivation beyond your interpretation: some error was faced in the past because too many arguments were issued to the compiler.

 I spent some time trying to find documentation on this 10000-input limit but couldn't find anything meaningful, so I left the implementation intact. @KKyang I believe that you implemented this feature, could you provide more context?

Comment (KKyang) [reply tensilelite/Tensile/TensileCreateLibrary.py]: https://github.com/ROCm/hipBLASLt/pull/415 this?

Comment (bstefanuk) [reply tensilelite/Tensile/TensileCreateLibrary.py]: That's correct.

**Comment by LunNova** [comment]
Might be worth zstd compressing the msgpack files too, they're pretty compressible. Here's an untested attempt at decompress support in case it's helpful:

```diff
diff --git a/Tensile/Source/lib/source/msgpack/MessagePack.cpp b/Tensile/Source/lib/source/msgpack/MessagePack.cpp
index de97929c..dbc397e0 100644
--- a/Tensile/Source/lib/source/msgpack/MessagePack.cpp
+++ b/Tensile/Source/lib/source/msgpack/MessagePack.cpp
@@ -28,6 +28,8 @@
 
 #include <Tensile/msgpack/Loading.hpp>
 
+#include <zstd.h>
+
 #include <fstream>
 
 namespace Tensile
@@ -86,6 +88,34 @@ namespace Tensile
                 return nullptr;
             }
 
+            // Check if the file is zstd compressed
+            char magic[4];
+            in.read(magic, 4);
+            bool isCompressed = (in.gcount() == 4 && magic[0] == '\x28' && magic[1] == '\xB5' && magic[2] == '\x2F' && magic[3] == '\xFD');
+            // Reset file pointer to the beginning
+            in.seekg(0, std::ios::beg);
+
+            if (isCompressed) {
+                // Decompress zstd file
+                std::vector<char> compressedData((std::istreambuf_iterator<char>(in)), std::istreambuf_iterator<char>());
+
+                size_t decompressedSize = ZSTD_getFrameContentSize(compressedData.data(), compressedData.size());
+                if (decompressedSize == ZSTD_CONTENTSIZE_ERROR || decompressedSize == ZSTD_CONTENTSIZE_UNKNOWN) {
+                    if(Debug::Instance().printDataInit())
+                        std::cout << "Error: Unable to determine decompressed size for " << filename << std::endl;
+                    return nullptr;
+                }
+
+                std::vector<char> decompressedData(decompressedSize);
+                size_t dSize = ZSTD_decompress(decompressedData.data(), decompressedSize, compressedData.data(), compressedData.size());
+                if (ZSTD_isError(dSize)) {
+                    if(Debug::Instance().printDataInit())
+                        std::cout << "Error: ZSTD decompression failed for " << filename << std::endl;
+                    return nullptr;
+                }
+
+                msgpack::unpack(result, decompressedData.data(), dSize);
+            } else {
             msgpack::unpacker unp;
             bool              finished_parsing;
             constexpr size_t  buffer_size = 1 << 19;
@@ -109,6 +139,7 @@ namespace Tensile
 
                 return nullptr;
             }
+            }
         }
         catch(std::runtime_error const& exc)
         {
```

**Comment by bstefanuk** [comment]
@LunNova Thanks for the code snippet. I've had this idea as well and have plans to implement it. However, for the scope of this PR we'll keep it to code object files and add the .dat file compression in another PR.

**Review by KKyang — DISMISSED** [review]

**Comment by bstefanuk** [comment]
@KKyang This change will increase build times by default in hipBLASLt. One option is to add a --release flag to the install script and TensileCreateLibrary, which will trigger compressed builds, which will take longer.

What is your opinion on the default behaviour, debug (fast, large binaries) or release (slow, small binaries)?

File: tensilelite/Tensile/TensileCreateLibrary.py
Comment (davidd-amd) [inline tensilelite/Tensile/TensileCreateLibrary.py:378]: This is a ***really good start*** to cleaning up the code used to generate binaries. In the future, I would like to consider something like:

```suggestion
    objectFiles += toolchain.compile(kernelFiles, outputPath)
    codeObjectFiles += toolchain.link(objectFiles, outputPath)
```

File: tensilelite/Tensile/Common.py
Comment (davidd-amd) [inline tensilelite/Tensile/Common.py:1625]: We should consider a new module for dealing with these types of operations.

File: tensilelite/Tensile/BuildCommands/AssemblyCommands.py
Comment (davidd-amd) [inline tensilelite/Tensile/BuildCommands/AssemblyCommands.py]: Right now it looks like we are unconditionally compressing. Do we plan to add an option to enable/disable as a part of this PR?

**Review by davidd-amd — APPROVED** [review]

File: tensilelite/Tensile/BuildCommands/AssemblyCommands.py
Comment (bstefanuk) [reply tensilelite/Tensile/BuildCommands/AssemblyCommands.py]: Done. Use the --no-compress flag.

**Review by jichangjichang — APPROVED** [review]
Please must get CQE's approval before merging. Thanks.

**Pull request merged**
